### PR TITLE
OAuth2 Client

### LINKED ISSUE #243: RestOnlyClient with Baerer Token Broken
# Summary
When using the RestOnlyClient with an Baerer Token, it throws 400: Bad Request

# Details
ASP.net / .net Core 2.0, Win10.

Its made shure that the Baerer Token works, because the normal DiscordClient is able to do it.
-> This may also be to the fact that DiscordClient does not trys to Get Guilds on start.

# Steps to reproduce
1. Get a Baerer Token
2. Create a Normal Discord Client -> See it work
3. Create a Rest Only Client -> Get your 400: Bad Request

### PR DESCRIPTION
**Pull request by HurricanKai**
# Summary
Fixes #243 It also Implements the OAuth2 Client Credentials Grant flow functionality. Adds Unit Tests for the Project.

# Details
Because i didnt wanted to wait for #243 i tryed it myself.
added Scope checking for all, it still contains some hacks but i hope its fine.
The Tests dont have to be there, i just like doing those.

# Changes proposed
* Client Credentials Grant
* OAuth2 Client, copyed from Rest-Only Client
     -> Now Merged with Rest Client
* OAuth2 Calls
* Tests
* Added Authorization Code Grant (as far as possible)
* automatic refresh (see comment)

# Notes
the Client Credentials Grant is not implemented for .Net Standart 1.1 because of unaviable Encoding.ASCII

I hope this will be helpfull 😄

### CODE REVIEW

**Comment by HurricanKai** [comment]
Added Authorization Code Grant.
Added Refreshing Tokens, this is not fully implemented because i dont know where to put the check.
There doesnt seem to be an event or something i coud hook into, but i dont wanna add a block like
```
if (CheckTokenRefresh())
	RefreshTokenAsync().Wait();
```
In every action

**Comment by NaamloosDT** [comment]
Please implement this into the rest client itself, as that was mostly intended for oauth2

**Comment by HurricanKai** [comment]
Moved To Rest as Requested by @NaamloosDT

**Comment by NaamloosDT** [comment]
lgtm, lemme try it out in a couple of mins

File: DSharpPlus/Utilities.cs [hunk 1]
```diff
@@ -53,6 +53,7 @@ static Utilities()
 
         internal static int CalculateIntegrity(int ping, DateTimeOffset timestamp, int heartbeat_interval)
         {
+
```
Comment (Emzi0767) [inline DSharpPlus/Utilities.cs]: ?

**Comment by HurricanKai** [comment]
Like this?

**Comment by NaamloosDT** [comment]
@nick-strohm can you finish your review? :^)

**Comment by HurricanKai** [comment]
Once the whatever you are doing right now, is finished, ill update the PR
(wont update every change, cause, well, unshure if i have to fix the same thing again ^^)

**Comment by uwx** [comment]
You should be safe to update your PR by now. If you need help merging, get in contact and I will help you.
Also, I think it would be nice to stay consistent and merge the two unit tests project, using MSTest instead of xunit, it should only be a few lines to change.

As a side note, watch out when doing unit tests on a proper client since that can hit the limit of connections pretty fast. I would say use `[ClassInitialize]` to connect and `[ClassCleanup]` to disconnect, that way there is only one connection per unit test run.

**Comment by NaamloosDT** [comment]
something something conflicting files

**Comment by HurricanKai** [comment]
Update to Current version. I hope its done right? I have not too much experience with Git ;d

**Comment by NaamloosDT** [comment]
did a little merge conflict fix before I start reviewing this (again)

**Comment by NaamloosDT** [comment]
Your unit tests seem to fail for me-

**Comment by HurricanKai** [comment]
Yeah, it prob. woud be, but i have no idea how to set that up lol

**Comment by uwx** [comment]
@HurricanKai It's been a while, will this PR ever be picked up? The rest client is still in a bit of a state of disrepair, not being up to date with the latest lib changes to the main client, and I think this would be a good starting point to get things back into shape

**Comment by uwx** [comment]
Well the only problems with the PR I can see are

* Branch has conflicts (due to the Optional serializer merge)
* Whitespace is all over the place (sometimes two spaces, sometimes 4 spaces, sometimes tabs)
* You're not [using EnumMember](https://github.com/DSharpPlus/DSharpPlus/pull/245/files#diff-925a13aae597cd20f626c3823e76ae69R1034)
* [DeserializeObject<dynamic>](https://github.com/DSharpPlus/DSharpPlus/pull/245/files#diff-925a13aae597cd20f626c3823e76ae69R991) can probably be optimized by using JObject (see the main lib for examples of that)
* [You've wrapped every method with a check](https://github.com/DSharpPlus/DSharpPlus/pull/245/files#diff-925a13aae597cd20f626c3823e76ae69R440), there surely has got to be a more terse way of doing it
* Exception messages are inconsistent with the rest of the lib
* You're still using SECRET.cs for test configuration, that could potentially break the CI build
* You're using XUnit, but the lib now uses MSTest, so it would be less ambiguous to move it all to MSTest (but if you don't do that I'll follow up with another PR for it)

File: DSharpPlus.Rest/DiscordRestClient.cs [hunk 2]
```diff
@@ -14,108 +19,233 @@ public class DiscordRestClient : BaseDiscordClient
...
+                    Optional<DefaultMessageNotifications> default_message_notifications, Optional<MfaLevel> mfa_level,
+                    Optional<ExplicitContentFilter> explicit_content_filter, Optional<ulong?> afk_channel_id,
+                    Optional<int> afk_timeout, Optional<string> iconb64, Optional<ulong> owner_id, Optional<string> splashb64,
+                    Optional<ulong?> systemChannelId, string reason)
+        {
+            if (scopes.Contains(Scope.bot))
+                return ApiClient.ModifyGuildAsync(guild_id, name, region, verification_level, default_message_notifications, mfa_level, explicit_content_filter, afk_channel_id, afk_timeout, iconb64,
+                    owner_id, splashb64, systemChannelId, reason);
+            else
+                throw new NotSupportedException("this is only Supported with the bot scope!");
```
Comment (Emzi0767) [inline DSharpPlus.Rest/DiscordRestClient.cs]: Exception messages should begin with a capital letter.

File: DSharpPlus.Rest/DiscordRestClient.cs [hunk 3]
```diff
@@ -429,5 +915,143 @@ public override void Dispose()
...
+        public void EnableRefresh(string refreshToken, string redirectUri, string clientId, string clientSecret)
+        {
+            this.RefreshToken = refreshToken;
+            this.redirectUri = redirectUri;
+            this.clientId = clientId;
+            this.clientSecret = clientSecret;
+        }
+
+        #region TokenGrants
+        //Not Included in NetStandart1.1 because of Encoding.ASCII
```
Comment (Emzi0767) [inline DSharpPlus.Rest/DiscordRestClient.cs]: Standard, not standart

File: DSharpPlus.Rest/Scope.cs [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+{
+#if !NETSTANDARD1_1
+    [Serializable]
+#endif
+    //[JsonConverter(typeof(StringEnumConverter))]
+    [Flags]
+    public enum Scope
+    {
+        [EnumMember(Value = "bot")]
+        bot,
```
Comment (Emzi0767) [inline DSharpPlus.Rest/Scope.cs]: Case member names properly.

File: DSharpPlus.Rest/DiscordRestClient.cs [hunk 5]
```diff
@@ -429,5 +915,143 @@ public override void Dispose()
...
+#if !NETSTANDARD1_1
+        //https://discordapp.com/developers/docs/topics/oauth2#client-credentials-grant
+        /// <summary>
+        /// This shoud only be used for Debug purposes
+        /// </summary>
+        public static async Task<DiscordRestClient> ClientCredentialsGrantAsync(string ClientId, string ClientSecret, Scope[] scopes, DiscordConfiguration config = null)
```
Comment (Emzi0767) [inline DSharpPlus.Rest/DiscordRestClient.cs]: Here, as well as in other instances, you should use enum flags, not arrays.

File: DSharpPlus.sln [hunk 6]
```diff
@@ -49,6 +49,10 @@ Project("{2150E333-8FDC-42A3-9474-1A3956D46DE8}") = ".github", ".github", "{430C
 		.github\PULL_REQUEST_TEMPLATE.md = .github\PULL_REQUEST_TEMPLATE.md
 	EndProjectSection
 EndProject
+Project("{9A19103F-16F7-4668-BE54-9A1E7A4F7556}") = "DSharpPlus.UnitTests", "DSharpPlus.UnitTests\DSharpPlus.UnitTests.csproj", "{7EAB8D4F-F17B-4594-8701-48CFC18E0882}"
+EndProject
+Project("{9A19103F-16F7-4668-BE54-9A1E7A4F7556}") = "DSharpPlus.OAuth", "DSharpPlus.OAuth\DSharpPlus.OAuth.csproj", "{1E53074B-1623-410F-A603-192AC7EF496F}"
```
Comment (Emzi0767) [inline DSharpPlus.sln]: Why a separate OAuth client when Rest project exists and it's there precisely for that reason?

File: DSharpPlus/Net/DiscordApiClient.cs [hunk 7]
```diff
@@ -1842,8 +1842,12 @@ private async Task<DiscordApplication> GetApplicationInfoAsync(string app_id)
         {
             var route = $"{Endpoints.OAUTH2}{Endpoints.APPLICATIONS}/:app_id";
             var bucket = this.Rest.GetBucket(RestRequestMethod.GET, route, new { app_id }, out var path);
-
-            var url = Utilities.GetApiUriFor(path);
+            //HACK: Replace with better solution
+            Uri url;
+            if (app_id == "@me" && Discord.Configuration.TokenType == TokenType.Bearer)
+              url = new Uri(string.Concat(Utilities.GetApiBaseUri().Replace("/v7", ""), path));
```
Comment (Emzi0767) [inline DSharpPlus/Net/DiscordApiClient.cs]: Your indentation settings are incorrect.

File: DSharpPlus/DiscordConfiguration.cs [hunk 8]
```diff
@@ -157,23 +157,26 @@ public UdpClientFactoryDelegate UdpClientFactory
...
-            this.ShardCount = other.ShardCount;
-            this.GatewayCompressionLevel = other.GatewayCompressionLevel;
-            this.MessageCacheSize = other.MessageCacheSize;
-            this.AutomaticGuildSync = other.AutomaticGuildSync;
-            this.WebSocketClientFactory = other.WebSocketClientFactory;
-            this.UdpClientFactory = other.UdpClientFactory;
-            this.Proxy = other.Proxy;
-            this.HttpTimeout = other.HttpTimeout;
-            this.ReconnectIndefinitely = other.ReconnectIndefinitely;
+            if (other != null)
```
Comment (Emzi0767) [inline DSharpPlus/DiscordConfiguration.cs]: You should check for null before passing a config, not within the constructor.

File: DSharpPlus.sln [hunk 9]
```diff
@@ -49,6 +49,10 @@ Project("{2150E333-8FDC-42A3-9474-1A3956D46DE8}") = ".github", ".github", "{430C
 		.github\PULL_REQUEST_TEMPLATE.md = .github\PULL_REQUEST_TEMPLATE.md
 	EndProjectSection
 EndProject
+Project("{9A19103F-16F7-4668-BE54-9A1E7A4F7556}") = "DSharpPlus.UnitTests", "DSharpPlus.UnitTests\DSharpPlus.UnitTests.csproj", "{7EAB8D4F-F17B-4594-8701-48CFC18E0882}"
```
Comment (Emzi0767) [inline DSharpPlus.sln]: Discord is not exactly unit testable.

**Comment by HurricanKai** [comment]
Arent the Intendation settings supposed to be 4 spaces?

**Comment by Kiritsu** [comment]
https://github.com/DSharpPlus/DSharpPlus/pull/245/files/a2a52cbfad1a7289afa686e7790a05a65810a2e6#diff-193b048ba2aab83b7964a19515ffaa88R1848
https://github.com/DSharpPlus/DSharpPlus/pull/245/files/a2a52cbfad1a7289afa686e7790a05a65810a2e6#diff-193b048ba2aab83b7964a19515ffaa88R1850

two spaces for each line here are missing

**Comment by NaamloosDT** [comment]
Closing due to inactivity

**Pull request closed without merging**